module/state_synchronization/execution_data_service.go

```diff
@@ -361,6 +371,7 @@ func (s *executionDataServiceImpl) getBlobs(ctx context.Context, cids []cid.Cid,
 // The returned error will be:
 // - MalformedDataError if some level of the blob tree cannot be properly deserialized
 // - BlobSizeLimitExceededError if any blob in the blob tree exceeds the maximum blob size
+// - ErrBlobTreeDepthExceeded if the blob tree exceeds the maximum depth
 // - BlobNotFoundError if some CID in the blob tree could not be found from the blobservice
 func (s *executionDataServiceImpl) Get(ctx context.Context, rootID flow.Identifier) (*ExecutionData, error) {
```

Comment:
We'd better to mention this call will also trigger to fetch execution data by the given rootID. And the call will be blocked until the full tree of CIDs has been fetched.

And now I have a question:
Is this method concurrent safe? If this method is called concurrently, would CIDs to be fetched multiple times or just once? Does execution data service maintain a `fetching` state to prevent fetching multiple times?

The scenario is that if the access node is answering multiple requests to return the execution data for a sealed height, but the requester has not downloaded it. Then execution data service's Get method might be called concurrently?

Reply:
AFAIK, there are not controls to prevent concurrent fetches of the same blob. The underlying db writes are using badger transactions and should concurrency safe, but concurrent gets would result in multiple downloads.

@smnzhu is that correct? Are you planning to address that in your storage upgrades?

Reply:
updated comment in 20e9b55